internal/statemachine/classic_states.go

```diff
@@ -70,20 +70,43 @@ func (stateMachine *StateMachine) parseImageDefinition() error {
 	}
 
 	// do custom validation for gadgetURL being required if gadget is not pre-built
-	if imageDefinition.Gadget.GadgetType != "prebuilt" && imageDefinition.Gadget.GadgetURL == "" {
-		jsonContext := gojsonschema.NewJsonContext("gadget_validation", nil)
-		errDetail := gojsonschema.ErrorDetails{
-			"key":   "gadget:type",
-			"value": imageDefinition.Gadget.GadgetType,
-		}
-		result.AddError(
-			imagedefinition.NewMissingURLError(
-				gojsonschema.NewJsonContext("missingURL", jsonContext),
-				52,
+	if imageDefinition.Gadget != nil {
+		if imageDefinition.Gadget.GadgetType != "prebuilt" && imageDefinition.Gadget.GadgetURL == "" {
+			jsonContext := gojsonschema.NewJsonContext("gadget_validation", nil)
+			errDetail := gojsonschema.ErrorDetails{
+				"key":   "gadget:type",
+				"value": imageDefinition.Gadget.GadgetType,
+			}
+			result.AddError(
+				imagedefinition.NewMissingURLError(
+					gojsonschema.NewJsonContext("missingURL", jsonContext),
+					52,
+					errDetail,
+				),
 				errDetail,
-			),
-			errDetail,
-		)
+			)
+		}
+	}
+
+	// don't allow any images to be created without a gadget
+	if imageDefinition.Gadget == nil {
+		if imageDefinition.Artifacts.Img == nil ||
+			imageDefinition.Artifacts.Iso == nil ||
+			imageDefinition.Artifacts.Qcow2 == nil {
```

Comment:
...it also might be just a brain fart from my side, so please correct me if I'm wrong: but this check here seems hm, the other way around? Since IIUC, we want to error out when there is NO gadget (`== nil`) AND there is an artifact defined that uses a gadget, right? So shouldn't the set of 'ors' be `!= nil`? To check that we error out if there's at least one artifact of img, iso or qcow2?

Reply:
Oh I see, it triggered on my test case because at least one of the image types was nil. Will fix.